### PR TITLE
Ret 3790 - Claimant views request from tribunal

### PR DESCRIPTION
**Pull request by marcjudge**
### JIRA link
https://tools.hmcts.net/jira/browse/RET-3790

### Change description
Claimant views request from tribunal and responds

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/controllers/helpers/ApplicationDetailsHelper.ts [hunk 1]
```diff
@@ -180,3 +188,180 @@ export const getTseApplicationDecisionDetails = (
...
+  respondCollection: TseRespondTypeItem[],
+  translations: AnyRecord,
+  accessToken: string,
+  res: Response
+): Promise<any> => {
+  const allResponses = [];
+
+  if (respondCollection) {
+    for (const response of respondCollection) {
+      if (response.value.from === Applicant.RESPONDENT || response.value.from === Applicant.CLAIMANT) {
```
Comment (Itay-Kirsh) [inline src/main/controllers/helpers/ApplicationDetailsHelper.ts:202]: it's either these or admin so you could do a switch and throw illegal state exception if not one of the three, not too bothered though, up to you.

File: src/main/controllers/helpers/ApplicationDetailsHelper.ts [hunk 2]
```diff
@@ -180,3 +188,180 @@ export const getTseApplicationDecisionDetails = (
...
+  }
+  return allResponses;
+};
+
+const getSupportingMaterialDownloadLink = async (
+  responseDoc: Document,
+  accessToken: string,
+  res: Response
+): Promise<string | void> => {
+  let responseDocDownload = undefined;
```
Comment (Itay-Kirsh) [inline src/main/controllers/helpers/ApplicationDetailsHelper.ts:356]: don't need to assign = undefined

File: src/main/controllers/helpers/DocumentHelpers.ts [hunk 3]
```diff
@@ -177,26 +190,27 @@ export function getSelectedAppResponseDocId(
...
-  let requestDocId = undefined;
-  if (requestDoc?.length) {
-    requestDocId = requestDoc.map(it => getDocId(it.value.uploadedDocument.document_url)).find(id => id === docId);
+export function isRequestDocId(req: AppRequest, docId: string): boolean {
+  if (
+    req.session.documentDownloadPage === PageUrls.TRIBUNAL_ORDER_OR_REQUEST_DETAILS ||
+    req.session.documentDownloadPage === PageUrls.GENERAL_CORRESPONDENCE_NOTIFICATION_DETAILS
+  ) {
+    const requestDoc = req.session?.userCase.selectedRequestOrOrder?.value.sendNotificationUploadDocument;
+    if (docId === requestDoc?.map(it => getDocId(it.value.uploadedDocument.document_url)).find(id => id === docId)) {
```
Comment (Itay-Kirsh) [inline src/main/controllers/helpers/DocumentHelpers.ts:199]: just return the condition

File: src/main/helper/ApiFormatter.ts [hunk 4]
```diff
@@ -481,7 +481,8 @@ export const setDocumentValues = (
 };
 
 export const getDocId = (url: string): string => {
-  return url.substring(url.lastIndexOf('/') + 1, url.length);
+  const x = url?.substring(url.lastIndexOf('/') + 1, url.length);
```
Comment (Itay-Kirsh) [inline src/main/helper/ApiFormatter.ts:484]: return x right away

Comment (Itay-Kirsh) [reply src/main/helper/ApiFormatter.ts:484]: Why this change?

**Review by Itay-Kirsh — CHANGES_REQUESTED** [review]

File: src/main/controllers/AttachmentController.ts [hunk 5]
```diff
@@ -26,85 +17,30 @@ export default class AttachmentController {
...
-    let appDocId;
-    let decisions;
-    let appDecisionDocId;
 
-    if (req.session.documentDownloadPage === PageUrls.RESPONDENT_APPLICATION_DETAILS) {
-      appDocId = getDocId(selectedApplicationDocUrl);
-      responseDocId = getResponseDocId(selectedApplication);
-      decisionDocId = getDecisionDocId(req, selectedApplication);
+    if (!docId) {
+      logger.info('no docId provided');
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/controllers/AttachmentController.ts:22]: Not sure how likely it is to happen, it was a test case though. When I ran the test I realised it checked every condition in the file before determining a docId was missing so I moved it to the top.
Added logger.warn.

File: src/main/controllers/AttachmentController.ts [hunk 6]
```diff
@@ -26,85 +17,30 @@ export default class AttachmentController {
...
+        logger.error(err.message);
         return res.redirect('/not-found');
       }
-      const document = await getCaseApi(req.session.user?.accessToken).getCaseDocument(docId);
-      res.setHeader('Content-Type', document.headers['content-type']);
-      res.status(200).send(Buffer.from(document.data, 'binary'));
-    } catch (err) {
-      logger.error(err.message);
+    } else {
+      logger.info('bad request parameter');
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/controllers/AttachmentController.ts:43]: Added

File: src/main/controllers/helpers/AdminNotificationHelper.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      if (request) {
+        appsWithTribunalOrderOrRequest.push(request);
+      }
+    }
+    return appsWithTribunalOrderOrRequest;
+  }
+};
+
+export const responseRequired = (adminRequest: AdminNotifcation): boolean => {
+  if (adminRequest?.isResponseRequired === 'Yes') {
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/controllers/helpers/AdminNotificationHelper.ts:24]: Added

File: src/main/helper/ApiFormatter.ts (hunk 4 above)
Comment (marcjudge) [reply src/main/helper/ApiFormatter.ts:484]: Mistakenly left in while debugging. Removed.

File: src/main/controllers/helpers/DocumentHelpers.ts (hunk 3 above)
Comment (marcjudge) [reply src/main/controllers/helpers/DocumentHelpers.ts:199]: Done

File: src/main/controllers/helpers/ApplicationDetailsHelper.ts (hunk 2 above)
Comment (marcjudge) [reply src/main/controllers/helpers/ApplicationDetailsHelper.ts:356]: Done

File: src/main/controllers/helpers/AdminNotificationHelper.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    return false;
+  }
+};
+
+export const getVisibleRequestFromAdmin = (
+  app: GenericTseApplicationTypeItem,
+  translations: AnyRecord,
+  languageParam: string
+): AdminNotifcation => {
+  if (app.value.respondCollection?.length > 0) {
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/controllers/helpers/AdminNotificationHelper.ts]: Updated the unnecessary check greater than zero.

**Review by DaveHmcts — DISMISSED** [review]

File: src/main/controllers/helpers/ApplicationDetailsHelper.ts [hunk 9]
```diff
@@ -180,3 +188,180 @@ export const getTseApplicationDecisionDetails = (
...
+
+export const getAllResponses = async (
+  respondCollection: TseRespondTypeItem[],
+  translations: AnyRecord,
+  accessToken: string,
+  res: Response
+): Promise<any> => {
+  const allResponses = [];
+
+  if (respondCollection) {
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/controllers/helpers/ApplicationDetailsHelper.ts]: Added

**Review by Itay-Kirsh — APPROVED** [review]

**Review by brendanmchugh — APPROVED** [review]

**Pull request merged**